src/App.tsx

```diff
@@ -1,24 +1,311 @@
+  updateTodoStatus,
+  updateTodoTitle,
+} from './api/todos';
+
+const USER_ID = 6160;
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<Errors>(Errors.NoError);
+  const [filteredTodos, setFilteredTodos] = useState<Todo[]>(todos);
+  const [selectedFilter, setSelectedFilter] = useState<Filters>(Filters.All);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [isNewTodoLoading, setIsNewTodoLoading] = useState(false);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const saveEditedTitle = (
+    todoId: number,
+    todoTitle: string,
+    setEditingTodoId: (value: number) => void,
+  ) => {
+    setLoadingTodoIds([...loadingTodoIds, todoId]);
+    updateTodoTitle(todoId, todoTitle)
+      .then(() => {
+        getTodos(USER_ID)
+          .then(todosFromServer => {
+            setTodos(todosFromServer);
+            setFilteredTodos(todosFromServer);
+            setLoadingTodoIds([]);
+            setEditingTodoId(0);
+          })
+          .catch(() => {
+            setError(Errors.CantUpdate);
+          });
+      });
+  };
+
+  const toggleTodoStatus = (todoId: number, completed: boolean) => {
+    updateTodoStatus(todoId, !completed)
+      .then(() => {
+        setLoadingTodoIds([...loadingTodoIds, todoId]);
+        getTodos(USER_ID)
+          .then(todosFromServer => {
+            setTodos(todosFromServer);
+            setFilteredTodos(todosFromServer);
+            setLoadingTodoIds([]);
+          })
+          .catch(() => {
+            setError(Errors.CantUpdate);
+          });
+      });
+  };
+
+  const getUpdatedTodos = (
+    todosToUpdate: Todo[],
+    everyTodoCompleted: boolean,
+    noTodoCompleted: boolean,
+    someTodosCompleted: boolean,
+  ) => {
+    const updatedTodos: Todo[] = [];
+
+    todosToUpdate.forEach(todo => {
+      let updatedTodo = { ...todo };
+
+      if (someTodosCompleted && !todo.completed) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
+
+      if (everyTodoCompleted || noTodoCompleted) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
+
+      updatedTodos.push(updatedTodo);
+    });
+
+    return updatedTodos;
+  };
+
+  const getLoadingTodoIds = (
+    todosToLoad: Todo[],
+    everyTodoCompleted: boolean,
+    noTodoCompleted: boolean,
+    someTodosCompleted: boolean,
+  ) => {
+    const loadingTodoIdsToAdd: number[] = [];
+
+    todosToLoad.forEach(todo => {
+      if (someTodosCompleted && !todo.completed) {
+        loadingTodoIdsToAdd.push(todo.id);
+      }
+
+      if (everyTodoCompleted || noTodoCompleted) {
+        loadingTodoIdsToAdd.push(todo.id);
+      }
```

Comment:
you can combine evaluations like in prev comment
```suggestion
      if ((someTodosCompleted && !todo.completed)
        || (everyTodoCompleted || noTodoCompleted)) {
        loadingTodoIdsToAdd.push(todo.id);
      }
```